### ISSUE TITLE
contains? not supported on type: clojure.lang.PersistentQueue

### ISSUE THREAD

**Issue by telekid** [dvlopt/dsim.cljc#4]
Not sure if this is a bug or if my mental model isn't quite right, but this code raises an exception when `a` and `b` are equal, which is surprising to me:

```clojure
(ns core
  (:require [dvlopt.dsim :as dsim]))

(def run
  (dsim/historic (dsim/ptime-engine)))

(defn held-note
  [duration]
  (fn [ctx]
    (dsim/e-conj ctx (dsim/queue (fn [ctx] (assoc-in ctx (dsim/path ctx) "on"))
                                 (dsim/wq-delay (dsim/rank+ duration))
                                 (fn [ctx] (assoc-in ctx (dsim/path ctx) "off"))))))

(defn midi-note [duration]
  (fn [ctx] (dsim/e-conj ctx
                         (dsim/ranks ctx)
                         (conj (dsim/path ctx) :note)
                         (held-note duration))))

(def a 10)
(def b 10)

(map #(dissoc % :dvlopt.dsim/events)
     (run (-> {}
              (dsim/e-conj [0] nil (midi-note a))
              (dsim/e-conj [b] nil (midi-note 20)))))

```

evaluating the final form results in:

```
1. Caused by java.lang.IllegalArgumentException
   contains? not supported on type: clojure.lang.PersistentQueue
```

However, if `a` and `b` are different (e.g. `(def a 9)`), it runs without failing. Any idea what I'm doing wrong? I'm wondering if part of my confusion stems from a poor understanding of the difference between `e-conj`, `e-assoc`, etc. It isn't entirely clear from the readme when I should be using the various `e-` fns.

**Comment by helins** [comment]
Hi! Sorry about the delay, I don't have a lot of time for open-source these days :)

There is a full description of the API on cljdoc, I don't know it is is apparent enough from the README, but here it is: https://cljdoc.org/d/dvlopt/dsim/1.0.0-beta1/doc/readme

So, your example fails because of what is going on with the `path`. I reckon it is not obvious and in the future I will make it throw with an explicit exception. Essentially, it fails because you try to schedule something deeper in the event tree (at `[:note]`) while there is already something scheduled at a lower path (at `nil`, the root of the event tree).

It ends up doing something similar to this, which also fails in a somewhat obscure way:

```clojure
(assoc-in {:a 42}
                [:a :b]
                :whatever)
=> java.lang.Long cannot be cast to clojure.lang.Associative
```

In general, it is always best to be explicit about the `path`, remember it is a simple but useful feature. Doing so greatly simplify things. For instance, in your example, you nest `e-conj` calls which make it hard to follow while I believe the same result is achieved with this simple implementation. I even mention the note number (60 = Middle C) in the path, I guess this is what you want:

```clojure
(require '[dvlopt.dsim :as dsim])

(def run
     (dsim/historic (dsim/ptime-engine)))

(defn held-note
  "Prepares a queue for holding a note."
  [duration]
  (dsim/queue (fn turn-on [ctx]
                (assoc-in ctx
                          (dsim/path ctx)
                          "on"))
              (dsim/wq-delay (dsim/rank+ duration))
              (fn turn-off [ctx]
                (assoc-in ctx
                          (dsim/path ctx)
                          "off"))))

(def a 10)
(def b 10)

(map #(dissoc %
              ::dsim/events)
     (run (-> {}
              (dsim/e-conj [0]
                           [:note 60]
                           (held-note a))
              (dsim/e-conj [b]
                           [:note 60]
                           (held-note 20)))))
```

I did fix a tiny little issue, try this code with version `1.0.0-beta1` ;)

**Comment by telekid** [comment]
Thanks for the response. It's taken a bit of hammock time, but things are all feeling a bit more intuitive now. Oh, and I appreciate the bugfix!
